lib/puppet/network/http/compression.rb

```diff
@@ -18,9 +18,10 @@ module Active
     # return an uncompressed body if the response has been
     # compressed
     def uncompress_body(response)
+      # TODO: is response.body already a BINARY string?
       case response['content-encoding']
       when 'gzip'
-        return Zlib::GzipReader.new(StringIO.new(response.body)).read
```

Comment:
Based on some debugging with @MosesMendoza on https://tickets.puppetlabs.com/browse/PUP-7251, we have determined that this change is not quite right.

When `response.body` is received, it is already an `Encoding::BINARY` string, which `StringIO` already understands when setting its encoding. Therefore additionally specifying `r+b` is actually meaningless.

The real problem here is that the value returned from `uncompress_body` is not in the same encoding as the value returned when compression is disabled.  With compression disabled, `response.body` is returned directly, and its encoding, as mentioned, is already `Encoding::BINARY`.

When using `Zlib::GzipReader`, the default encoding of the returned string is actually `Encoding.default_external` - which could be an array of things varying on the platform, like `Encoding::CP932` (in Japan) or `Encoding::ISO8859_1` (in Europe).  Fortunately, the `Zlib::GzipReader.new` initializer can accept options that specify `:encoding => Encoding::BINARY` so that we continue to return binary strings for PSON when compression is on, as we're doing when compression is off.

Therefore as part of [PUP-7251](https://tickets.puppetlabs.com/browse/PUP-7251) the fix for this will probably be:

```ruby
return Zlib::GzipReader.new(StringIO.new(response.body, 'r+b'), :encoding => Encoding::BINARY).read
```